### PR TITLE
ref: sentry-core changes for console app

### PR DESCRIPTION
**Pull request by marandaneto**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [X] Refactoring

## :scroll: Description
ref: sentry-core changes for console app

## :bulb: Motivation and Context
it's not possible to use sentry-core standalone, this is a first PR that makes things a bit easier.

## :pencil: Checklist

- [X] I reviewed submitted code
- [X] I added tests to verify changes
- [X] All tests passing

### CODE REVIEW

File: sentry-core/src/main/java/io/sentry/core/Sentry.java [hunk 1]
```diff
@@ -160,9 +154,65 @@ public static void init(
...
+    logger.log(SentryLevel.INFO, "Initializing SDK with DSN: '%s'", options.getDsn());
+
+    if (options.getSentryClientName() == null || options.getSentryClientName().isEmpty()) {
+      options.setSentryClientName("sentry.java/2.2.0"); // TODO: get from build script or CI envs
+    }
+
+    // TODO: read values from conf file eg Manifest
+    // eg release and distinctId
+
+    if (options.getEnvelopeReader() instanceof NoOpEnvelopeReader) {
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/Sentry.java:185]: What about we jsut add `EnvelopeReader` by default instead of `NoOpEnvelopeReader`?

Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/Sentry.java:185]: wfm, actually I even wanted to do this with other objects that should not be replaced anyway, we can discuss what it's make sense to be the object by default and not the NoOp version of it

File: sentry-core/src/main/java/io/sentry/core/transport/HttpTransport.java [hunk 2]
```diff
@@ -166,6 +167,12 @@ public HttpTransport(
     try (final OutputStream outputStream = connection.getOutputStream();
         final GZIPOutputStream gzip = new GZIPOutputStream(outputStream);
         final Writer writer = new BufferedWriter(new OutputStreamWriter(gzip, UTF_8))) {
+
+      // TODO: just for debugging, remove it
+      StringWriter stringWriter = new StringWriter();
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/transport/HttpTransport.java:172]: yep, we do something similar for logging the error payload from the API, if there's an error, sounds good.

File: sentry-core/src/main/java/io/sentry/core/Sentry.java (hunk 1 above)
Comment (bruno-garcia) [reply sentry-core/src/main/java/io/sentry/core/Sentry.java:185]: Fair. Let's aim at: If it's likely the default implementation will be used (or we most certainly need one that is not NoOp) we should initialize with our default instance.

File: sentry-core/src/main/java/io/sentry/core/SentryOptions.java [hunk 3]
```diff
@@ -95,6 +102,11 @@
   /** The sessions dir. size for capping the number of envelopes Default is 100 */
   private int sessionsDirSize = 100;
 
+  private int maxQueueSize = cacheDirSize + sessionsDirSize;
+  // TODO: revert, this is necessary because start/end session + events will be higher than the
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/SentryOptions.java:106]: I'm not sure I follow this comment. This maxQueueSize is not a single valid used by any file queue?
btw there are some changes that we'll be doing which will mean a single queue of envelopes to be submitted. That's to address the session updates vs events sync. So going forward we'll likely just merge and have a single folder

File: sentry-sample/src/main/java/io/sentry/sample/MainActivity.java [hunk 4]
```diff
@@ -58,6 +60,56 @@ protected void onCreate(Bundle savedInstanceState) {
           }
         });
 
+    binding.sessions.setOnClickListener(
+        view -> {
+          for (int i = 0; i < 100; i++) {
+            System.out.println("session start: i = " + i);
```
Comment (bruno-garcia) [inline sentry-sample/src/main/java/io/sentry/sample/MainActivity.java:66]: H: Is this change related to the session data generator? Is this going to land in this PR?

File: sentry-android-core/src/main/java/io/sentry/android/core/AndroidOptionsInitializer.java [hunk 5]
```diff
@@ -98,37 +96,32 @@ static void init(
...
       final @NotNull IBuildInfoProvider buildInfoProvider,
       final @NotNull ILoadClass loadClass) {
 
     // Integrations are registered in the same order. NDK before adding Watch outbox,
     // because sentry-native move files around and we don't want to watch that.
     final Class<?> sentryNdkClass = loadNdkIfAvailable(options, buildInfoProvider, loadClass);
     options.addIntegration(new NdkIntegration(sentryNdkClass));
-    options.addIntegration(EnvelopeFileObserverIntegration.getOutboxFileObserver(envelopeReader));
+
+    // TODO: do we need this for Java backend? dont think so
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidOptionsInitializer.java:123]: we can't, `EnvelopeFileObserverIntegration` uses `EnvelopeFileObserver`, which extends `android.os.FileObserver`, it's not pure Java, we could have an interface, but still, it won't be working anyway.

File: sentry-android-core/src/main/java/io/sentry/android/core/NdkIntegration.java [hunk 6]
```diff
@@ -26,6 +26,12 @@ public final void register(final @NotNull IHub hub, final @NotNull SentryOptions
     Objects.requireNonNull(hub, "Hub is required");
     Objects.requireNonNull(options, "SentryOptions is required");
 
+    final String cachedDir = options.getCacheDirPath();
+    if (cachedDir == null || cachedDir.isEmpty()) {
+      options.getLogger().log(SentryLevel.WARNING, "No cache dir path is defined in options.");
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/NdkIntegration.java:31]: it depends the point of view, as we add the integration and the flag is enabled by default, it'd make sense to be an error, but having a caching folder is optional anyway, so I'd say it's a warning, if the user disables `isEnabledSdk`, then this won't be even ran

File: sentry-core/src/main/java/io/sentry/core/DiagnosticLogger.java [hunk 7]
```diff
@@ -29,8 +29,8 @@ public DiagnosticLogger(final SentryOptions options, final @Nullable ILogger log
    * @return True if a log message would be recorded for the level. Otherwise false.
    */
   public boolean isEnabled(final @Nullable SentryLevel level) {
-    SentryLevel diagLevel = options.getDiagnosticLevel();
-    if (level == null || diagLevel == null) {
+    final SentryLevel diagLevel = options.getDiagnosticLevel();
+    if (level == null) {
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/DiagnosticLogger.java:33]: it does for the level that comes as a param, but the `diagnosticLevel` is never null, if one sets null, its gonna take the DEFAULT_DIAGNOSTIC_LEVEL anyway

File: sentry-core/src/main/java/io/sentry/core/SentryOptions.java [hunk 8]
```diff
@@ -68,6 +73,8 @@
   /** Serializer interface to serialize/deserialize json events */
   private @NotNull ISerializer serializer = NoOpSerializer.getInstance();
 
+  private @NotNull IEnvelopeReader envelopeReader = NoOpEnvelopeReader.getInstance();
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/SentryOptions.java:76]: tbh I dont see the need of having a `NoOpEnvelopeReader` but all the other fields have it too, so I decided to create it, happy with creating an `EnvelopeReader` instead

File: sentry-core/src/main/java/io/sentry/core/SentryOptions.java (hunk 3 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/SentryOptions.java:106]: our thread/transport queue is 10 items and it drops all the next threads that will try to be enqueued.
as I was doing a test with 100 sessions in a for loop, without a proper delay between them, I needed to increase the queue size.
the comment is just to UNDO it, this most probably is not a real use case, and the queue size is a param anyway, prople can increase if they want to

File: sentry-sample/src/main/java/io/sentry/sample/MainActivity.java (hunk 4 above)
Comment (marandaneto) [reply sentry-sample/src/main/java/io/sentry/sample/MainActivity.java:66]: related to session data generator, yes, land in this PR, no

**Comment by marandaneto** [comment]
> It looks like the right direction. Are we far from dropping the draft flag?
> It's getting pretty daunting to review this, I hope we can merge soon

interested to learn why it's getting `daunting`.
nothing changed from your last review, there are no new commits.
basically the next step is to clean up the code and fix tests, work on your comments if any, nothing more to be changed or added, only removed.

File: sentry-android-core/src/main/java/io/sentry/android/core/AndroidOptionsInitializer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bruno-garcia) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidOptionsInitializer.java:123]: That's a good point. We'd need to have an interface and a pure Java implementation that uses a background thread and polls the file system for files every X seconds. Let's create a task for that

File: sentry-sample/src/main/java/io/sentry/sample/MainActivity.java (hunk 4 above)
Comment (bruno-garcia) [reply sentry-sample/src/main/java/io/sentry/sample/MainActivity.java:66]: why is this in the diff?

File: sentry-sample/src/main/AndroidManifest.xml [hunk 9]
```diff
@@ -3,88 +3,100 @@
...
+        <!--    <meta-data android:name="io.sentry.uncaught-exception-handler.enable" android:value="false" />-->
 
-    <!--    how to disable the activity lifecycle breadcrumbs integration-->
-    <!--    <meta-data android:name="io.sentry.breadcrumbs.activity-lifecycle" android:value="false" />-->
+        <!--    how to disable the activity lifecycle breadcrumbs integration-->
+        <!--    <meta-data android:name="io.sentry.breadcrumbs.activity-lifecycle" android:value="false" />-->
 
-    <!--    how to disable the app lifecycle breadcrumbs integration-->
-    <!--    <meta-data android:name="io.sentry.breadcrumbs.app-lifecycle" android:value="false" />-->
+        <!--    how to disable the app lifecycle breadcrumbs integration-->
```
Comment (bruno-garcia) [inline sentry-sample/src/main/AndroidManifest.xml:87]: this is just a linter thing? I dunno what I'm reviewing here

File: sentry-sample/src/main/res/layout/activity_main.xml [hunk 10]
```diff
@@ -1,43 +1,56 @@
 <?xml version="1.0" encoding="utf-8"?>
 <LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
-              xmlns:tools="http://schemas.android.com/tools"
```
Comment (bruno-garcia) [inline sentry-sample/src/main/res/layout/activity_main.xml:3]: What changed here? Not sure what to review

File: sentry-core/src/main/java/io/sentry/core/SentryOptions.java (hunk 3 above)
Comment (bruno-garcia) [reply sentry-core/src/main/java/io/sentry/core/SentryOptions.java:106]: Got it. Makes sense now, thanks for clarifying.

File: sentry-core/src/main/java/io/sentry/core/SentryOptions.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (bruno-garcia) [reply sentry-core/src/main/java/io/sentry/core/SentryOptions.java:76]: sounds good

File: sentry-core/src/main/java/io/sentry/core/Sentry.java [hunk 11]
```diff
@@ -294,7 +346,7 @@ public static void setTransaction(@Nullable String transaction) {
    *
    * @param user the user
    */
-  public static void setUser(@Nullable User user) {
+  public static void setUser(final @Nullable User user) {
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/Sentry.java:349]: Super hard to review this PR with so much unrelated changes like this

File: sentry-core/src/main/java/io/sentry/core/Sentry.java [hunk 12]
```diff
@@ -238,7 +289,8 @@ public static synchronized void close() {
    * @param breadcrumb the breadcrumb
    * @param hint SDK specific but provides high level information about the origin of the event
    */
-  public static void addBreadcrumb(@NotNull Breadcrumb breadcrumb, @Nullable Object hint) {
+  public static void addBreadcrumb(
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/Sentry.java:292]: Super hard to review this PR with so much unrelated changes like this

File: sentry-core/src/main/java/io/sentry/core/Sentry.java [hunk 13]
```diff
@@ -86,12 +90,12 @@ public static void init() {
    * @throws InvocationTargetException the InvocationTargetException
    */
   public static <T extends SentryOptions> void init(
-      @NotNull OptionsContainer<T> clazz,
-      @NotNull OptionsConfiguration<T> optionsConfiguration,
-      boolean globalHubMode)
+      final @NotNull OptionsContainer<T> clazz,
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/Sentry.java:93]: Super hard to review this PR with so much unrelated changes like this

File: sentry-core/src/main/java/io/sentry/core/SendCachedEventFireAndForgetIntegration.java [hunk 14]
```diff
@@ -12,32 +15,59 @@
...
+        directoryProcessor.processDirectory(dirFile);
+
+        logger.log(SentryLevel.DEBUG, "Finished processing cached files from %s", dirPath);
+      };
+    }
   }
 
-  public SendCachedEventFireAndForgetIntegration(SendFireAndForgetFactory factory) {
-    this.factory = factory;
+  public SendCachedEventFireAndForgetIntegration(final @NotNull SendFireAndForgetFactory factory) {
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/SendCachedEventFireAndForgetIntegration.java:49]: Super hard to review this PR with so much unrelated changes like this

File: sentry-core/src/main/java/io/sentry/core/EnvelopeSender.java [hunk 15]
```diff
@@ -214,8 +214,6 @@ private void processEnvelope(final @NotNull SentryEnvelope envelope, final @Null
               items);
           break;
         }
-      } else {
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/EnvelopeSender.java:217]: Is this needed to changing to sentry-core? Or just another refactor?

File: sentry-core/src/main/java/io/sentry/core/AsyncConnectionFactory.java [hunk 16]
```diff
@@ -17,7 +17,7 @@ public static AsyncConnection create(
         options.getTransportGate(),
         eventCache,
         sessionCache,
-        options.getCacheDirSize(),
+        options.getMaxQueueSize(),
```
Comment (bruno-garcia) [inline sentry-core/src/main/java/io/sentry/core/AsyncConnectionFactory.java:20]: M: MaxQueueSize indicates the number of items in-memory that we can hold before dropping. Is this what this is?
The original thing meant max items in the file system, right?

File: sentry-android-core/src/main/java/io/sentry/android/core/NdkIntegration.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (bruno-garcia) [reply sentry-android-core/src/main/java/io/sentry/android/core/NdkIntegration.java:31]: The cache folder is optional if you don't use the NDK integration. Otherwise it'll be broken so I'd say it makes sense to log error if we have NDK + no cache path.

File: sentry-android-core/src/main/java/io/sentry/android/core/EnvelopeFileObserverIntegration.java [hunk 17]
```diff
@@ -15,21 +15,20 @@
...
+  public static @NotNull EnvelopeFileObserverIntegration getOutboxFileObserver() {
+    return new OutboxEnvelopeFileObserverIntegration();
   }
 
   @Override
-  public final void register(IHub hub, SentryOptions options) {
-    ILogger logger = options.getLogger();
-    String path = getPath(options);
+  public final void register(final @NotNull IHub hub, final @NotNull SentryOptions options) {
+    Objects.requireNonNull(hub, "Hub is required");
```
Comment (bruno-garcia) [inline sentry-android-core/src/main/java/io/sentry/android/core/EnvelopeFileObserverIntegration.java:26]: Hard to review this since it's just refactoring unrelated to the change

File: sentry-android-core/src/main/java/io/sentry/android/core/AndroidOptionsInitializer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bruno-garcia) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidOptionsInitializer.java:123]: We only use this for NDK right? So doesn't matter I guess

[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/AndroidOptionsInitializer.java:123]: also for React native

File: sentry-android-core/src/main/java/io/sentry/android/core/NdkIntegration.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/NdkIntegration.java:31]: it's not optional, this PR makes it optional.

File: sentry-sample/src/main/java/io/sentry/sample/MainActivity.java (hunk 4 above)
Comment (marandaneto) [reply sentry-sample/src/main/java/io/sentry/sample/MainActivity.java:66]: aka Draft PR, we agreed to lift a draft PR to share the knowledge of the session data generator (how we did it), this was asked by you btw on a daily standup.

File: sentry-sample/src/main/AndroidManifest.xml (hunk 9 above)
Comment (marandaneto) [reply sentry-sample/src/main/AndroidManifest.xml:87]: probably changed by the IDE automatically and will be reverted, Draft PR

File: sentry-sample/src/main/res/layout/activity_main.xml (hunk 10 above)
Comment (marandaneto) [reply sentry-sample/src/main/res/layout/activity_main.xml:3]: probably changed by the IDE automatically and will be reverted, Draft PR

File: sentry-core/src/main/java/io/sentry/core/Sentry.java (hunk 11 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/Sentry.java:349]: true, I should not have changed this, but it's a Draft right

File: sentry-core/src/main/java/io/sentry/core/Sentry.java (hunk 12 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/Sentry.java:292]: true, I should not have changed this, but it's a Draft right

File: sentry-core/src/main/java/io/sentry/core/Sentry.java (hunk 13 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/Sentry.java:93]: true, I should not have changed this, but it's a Draft right

File: sentry-core/src/main/java/io/sentry/core/SendCachedEventFireAndForgetIntegration.java (hunk 14 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/SendCachedEventFireAndForgetIntegration.java:49]: this class needed to be changed, so it made sense to "improve" the missing finals and annotations IMO.
this class has more changes than this and adding NotNull helps with interoperability with Kotlin tests

File: sentry-core/src/main/java/io/sentry/core/EnvelopeSender.java (hunk 15 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/EnvelopeSender.java:217]: this class needed to be changed, so it made sense to "clean up" dead code, it was a useless logging

File: sentry-core/src/main/java/io/sentry/core/AsyncConnectionFactory.java (hunk 16 above)
Comment (marandaneto) [reply sentry-core/src/main/java/io/sentry/core/AsyncConnectionFactory.java:20]: yes, exactly, I believe with session + events, using the CacheDirSize doesn't make sense anymore and we should have a specific option (MaxQueueSize) for that

File: sentry-android-core/src/main/java/io/sentry/android/core/EnvelopeFileObserverIntegration.java (hunk 17 above)
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/EnvelopeFileObserverIntegration.java:26]: this class needed to be changed, so it made sense to "improve" the missing finals and annotations IMO.
this class has more changes than this and adding NotNull helps with interoperability with Kotlin tests

**Comment by marandaneto** [comment]
> This is just a draft and it looks huge:
> 
> image
> image
> 
> Probably just 20% of that is actually relevant to the change. Please let's avoid mixing a bunch of refactor in PRs that add a feature or change a core part of the SDK.

while I agree with the comment, 

> This is just a draft and it looks huge:
> 
> image
> image
> 
> Probably just 20% of that is actually relevant to the change. Please let's avoid mixing a bunch of refactor in PRs that add a feature or change a core part of the SDK.

while I agree with this statement, we agreed to lift a draft PR to share the knowledge of the session data generator (how we did it), this was asked by you btw on a daily standup.

the only code that actually needed to land in the master branch was #474 which is a separate PR, tested, and merged.

all these changes came for free because I actually needed then to make sentry-core to work standalone.
I'm happy to close this PR and make a new one.
This PR is specifically for sending session data to Sentry and validating its APIs, maybe the title of this PR is misleading, but this was stated in the standup.

Let's please not put a lot of useless comments in a Draft PR and focus on the specific bits mentioned in the daily call, which is the changes from sentry-core that actually is gonna be needed for the new sentry-java, all the rest will be discarded, but if I didn't push it, you won't know how I generated the session data.

I'm trying to improve things here, and trying to reuse some work for the future sentry-java, this PR is just early feedback of "how I would fix sentry-core to be usable in sentry-java", so adding negative/useless comments just makes it harder to continue.

happy to discuss offline as well.

File: sentry-sample/src/main/java/io/sentry/sample/MainActivity.java (hunk 4 above)
Comment (bruno-garcia) [reply sentry-sample/src/main/java/io/sentry/sample/MainActivity.java:66]: It's fine that it's draft, but it was unclear if this is planned to be merged or not.
The "refactors" like `@Null` and `final` I don't expect would be deleted once the PR drops the Draft, or would it?

Sorry if it was too detailed for a Draft PR. I won't review this a third time as Draft then. I'll wait until we can get closer to review it.

**Comment by bruno-garcia** [comment]
> while I agree with this statement, we agreed to lift a draft PR to share the knowledge of the session data generator (how we did it), this was asked by you btw on a daily standup.

I wasn't aware they'd come into the same PR. I was genuinely askign why that code is in this PR, since I didn't know if the plan was to merge both things or not, at the same time. Reason why I suggested we don't do that.

> I'm trying to improve things here, and trying to reuse some work for the future sentry-java, this PR is just early feedback of "how I would fix sentry-core to be usable in sentry-java", so adding negative/useless comments just makes it harder to continue.

I understand and appreciate the work you put into this and the constant improvement of things. I wasn't trying to add any "negative/useless" comment, simply reiterating what we discussed in several meetings, to try to avoid doing different types of changes (refactor, different features, etc) in a single PR since it can get really difficult to review things.

> all these changes came for free because I actually needed them to make sentry-core to work standalone.
I'm happy to close this PR and make a new one.
This PR is specifically for sending session data to Sentry and validating its APIs, maybe the title of this PR is misleading, but this was stated in the standup.

The title is what made it confusing because I thought it was simply the change, as the title says _sentry-core changes for console app_ so all the other code seem not to fit here.

We can catch up offline, no problem. I'm OOO until Tuesday though, sorry.

File: sentry-sample/src/main/java/io/sentry/sample/MainActivity.java (hunk 4 above)
Comment (marandaneto) [reply sentry-sample/src/main/java/io/sentry/sample/MainActivity.java:66]: clarified offline

File: sentry-android-core/src/main/java/io/sentry/android/core/SentryAndroidOptions.java [hunk 18]
```diff
@@ -32,6 +33,11 @@
   /** Enable or disable automatic breadcrumbs for App Components Using ComponentCallbacks */
   private boolean enableAppComponentBreadcrumbs = true;
 
+  public SentryAndroidOptions() {
+    setSentryClientName(BuildConfig.SENTRY_CLIENT_NAME + "/" + BuildConfig.VERSION_NAME);
+    setSdkInfo(SdkInfo.createSdkInfo(BuildConfig.SENTRY_CLIENT_NAME, BuildConfig.VERSION_NAME));
```
Comment (bruno-garcia) [inline sentry-android-core/src/main/java/io/sentry/android/core/SentryAndroidOptions.java:38]: This should be an override to `SentryOptions` ctor which sets it to `sentry.java`
